File: 2020/day22/day22.rs
//! [Day 22: Crab Combat](https://adventofcode.com/2020/day/22)

use rustc_hash::FxHashSet;

/// Maximum length a deck.
/// 64 is faster than the minimum required value (50).
const SIZE: usize = 64;

#[derive(Clone)]
struct Deck {
    cards: [u8; SIZE],
    length: usize,
    bottom: usize,
}

impl Deck {
    #[inline]
    const fn new() -> Self {
        Self {
            cards: [0u8; SIZE],
            length: 0,
            bottom: 0,
        }
    }

    fn parse(cards: &str) -> Self {
        let mut deck = Self::new();
        for c in cards.lines().skip(1).map(|s| s.parse().unwrap()) {
            deck.push(c);
        }
        deck
    }

    #[inline]
    const fn is_empty(&self) -> bool {
        self.length == 0
    }

    #[inline]
    fn push(&mut self, card: u8) {
        self.cards[(self.bottom + self.length) % SIZE] = card;
        self.length += 1;
    }

    #[inline]
    fn pop(&mut self) -> u8 {
        let card = self.cards[self.bottom];
        self.bottom = (self.bottom + 1) % SIZE;
        self.length -= 1;
        card
    }

    #[inline]
    fn extend(&mut self, other: &Self) {
        for i in 0..other.length {
            self.cards[(self.bottom + self.length) % SIZE] = other.cards[(other.bottom + i) % SIZE];
            self.length += 1;
        }
    }

    fn score(&self) -> u32 {
        let mut idx = (self.bottom + self.length) % SIZE;
        let mut rank = 0;
        let mut s = 0;
        for _ in 0..self.length {
            idx = (idx + SIZE - 1) % SIZE;
            rank += 1;
            s += rank * u32::from(self.cards[idx]);
        }
        s
    }

    #[inline]
    fn take(&self, n: usize) -> Self {
        let mut new_deck = self.clone();
        new_deck.length = n;
        new_deck
    }
}

/// Make a hash with the two decks.
/// The hash is made with cards of the first deck, then 0, then the cards of the second deck.
fn make_hash(deck1: &Deck, deck2: &Deck) -> [u8; 64] {
    let mut hash = [0u8; 64];

    #[allow(clippy::needless_range_loop)]
    for i in 0..deck1.length {
        hash[i] = deck1.cards[(deck1.bottom + i) % SIZE];
    }

    for i in 0..deck2.length {
        hash[i + deck1.length + 1] = deck2.cards[(deck2.bottom + i) % SIZE];
    }

    hash
}

fn recursive_combat(deck1: &Deck, deck2: &Deck, recursive: bool) -> (Deck, Deck) {
    let mut seen = FxHashSet::default();

    let mut deck1 = deck1.clone();
    let mut deck2 = deck2.clone();

    while !deck1.is_empty() && !deck2.is_empty() {
        if recursive && !seen.insert(make_hash(&deck1, &deck2)) {
            deck1.extend(&deck2);
            return (deck1, Deck::new());
        }

        let card1 = deck1.pop();
        let card2 = deck2.pop();

        let deck1_win = if recursive
            && deck1.length >= usize::from(card1)
            && deck2.length >= usize::from(card2)
        {
            recursive_combat(
                &deck1.take(usize::from(card1)),
                &deck2.take(usize::from(card2)),
                true,
            )
            .1 //  deck2
            .is_empty()
        } else {
            card1 > card2
        };

        if deck1_win {
            deck1.push(card1);
            deck1.push(card2);
        } else {
            deck2.push(card2);
            deck2.push(card1);
        }
    }

    (deck1, deck2)
}

struct Puzzle {
    deck1: Deck,
    deck2: Deck,
}

impl Puzzle {
    /// Initialize from the puzzle input.
    fn new(data: &str) -> Self {
        let (cards1, cards2) = data.split_once("\n\n").unwrap();

        Self {
            deck1: Deck::parse(cards1),
            deck2: Deck::parse(cards2),
        }
    }

    /// Solve part one.
    fn part1(&self) -> u32 {
        let (win1, win2) = recursive_combat(&self.deck1, &self.deck2, false);
        win1.score() + win2.score()
    }

    /// Solve part two.
    fn part2(&self) -> u32 {
        let (win1, win2) = recursive_combat(&self.deck1, &self.deck2, true);
        win1.score() + win2.score()
    }
}

fn main() {
    let args = aoc::parse_args();
    let puzzle = Puzzle::new(&args.input);
    println!("{}", puzzle.part1());
    println!("{}", puzzle.part2());
}

/// Test from puzzle input
#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn part1() {
        let data = aoc::load_input_data("test.txt");
        let puzzle = Puzzle::new(&data);
        assert_eq!(puzzle.part1(), 306);
    }

    #[test]
    fn part2() {
        let data = aoc::load_input_data("test.txt");
        let puzzle = Puzzle::new(&data);
        assert_eq!(puzzle.part2(), 291);
    }
}

File: src/year2020/day22/day22.py
#!/usr/bin/env python3
# [Day 22: Crab Combat](https://adventofcode.com/2020/day/22)

import atexit
import sys
import time
from pathlib import Path

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


def recursive_combat(deck1, deck2, recursive=False):
    seen = set()

    while deck1 and deck2:
        state = (tuple(deck1), tuple(deck2))
        if recursive and state in seen:
            return deck1 + deck2, []
        seen.add(state)

        card1, card2 = deck1.pop(0), deck2.pop(0)

        if recursive and len(deck1) >= card1 and len(deck2) >= card2:
            win1, _ = recursive_combat(deck1[:card1], deck2[:card2], True)
            if len(win1) > 0:
                deck1.extend((card1, card2))
            else:
                deck2.extend((card2, card1))
        else:
            if card1 > card2:
                deck1.extend((card1, card2))
            else:
                deck2.extend((card2, card1))

    return deck1, deck2


# part 1/2
for part in (False, True):
    deck1, deck2 = data.split("\n\n")
    deck1 = list(map(int, deck1.splitlines()[1:]))
    deck2 = list(map(int, deck2.splitlines()[1:]))

    win1, win2 = recursive_combat(deck1, deck2, part)
    winning = win1 if len(win1) > 0 else win2
    print(sum(s * c for s, c in enumerate(reversed(winning), 1)))
